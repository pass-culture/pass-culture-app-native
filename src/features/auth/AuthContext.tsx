import NetInfo from '@react-native-community/netinfo'
import React, { memo, useCallback, useContext, useEffect, useMemo, useState } from 'react'
import { useQueryClient } from 'react-query'

import { api } from 'api/api'
import { refreshAccessToken } from 'api/apiHelpers'
import { SigninResponse } from 'api/gen'
import { useResetContexts } from 'features/auth/useResetContexts'
import { analytics, LoginRoutineMethod } from 'libs/analytics'
import { useAppStateChange } from 'libs/appState'
import { getAccessTokenStatus, getUserIdFromAccesstoken } from 'libs/jwt'
import { clearRefreshToken, saveRefreshToken } from 'libs/keychain'
import { eventMonitoring } from 'libs/monitoring'
import { QueryKeys } from 'libs/queryKeys'
import { BatchUser } from 'libs/react-native-batch'
import { storage } from 'libs/storage'

export interface IAuthContext {
  isLoggedIn: boolean
  setIsLoggedIn: (isLoggedIn: boolean) => void
}

const connectUserToBatchAndFirebase = (accessToken: string | null) => {
  if (!accessToken) return
  const userId = getUserIdFromAccesstoken(accessToken)
  if (userId) {
    BatchUser.editor().setIdentifier(userId.toString()).save()
    analytics.setUserId(userId)
    eventMonitoring.setUser({ id: userId.toString() })
  }
}

export const AuthContext = React.createContext<IAuthContext>({
  isLoggedIn: false,
  setIsLoggedIn: () => undefined,
})

export function useAuthContext(): IAuthContext {
  return useContext(AuthContext)
}

export const AuthWrapper = memo(function AuthWrapper({ children }: { children: JSX.Element }) {
  const [loading, setLoading] = useState(true)
  const [isLoggedIn, setIsLoggedIn] = useState(false)

  const readTokenAndConnectUser = useCallback(async () => {
    try {
      let accessToken = await storage.readString('access_token')
      const { isConnected } = await NetInfo.fetch()

      if (!!accessToken && !isConnected) {
        setIsLoggedIn(true)
        return
      }

      if (getAccessTokenStatus(accessToken) === 'expired') {
        // refreshAccessToken calls the backend to get a new access token
        // and also saves it to the storage
        accessToken = await refreshAccessToken(api)
      }

      if (getAccessTokenStatus(accessToken) === 'valid') {
        setIsLoggedIn(true)
        connectUserToBatchAndFirebase(accessToken)
      }
    } catch (err) {
      eventMonitoring.captureException(err)
      setIsLoggedIn(false)
    } finally {
      setLoading(false)
    }
  }, [])

  useEffect(() => {
    readTokenAndConnectUser()
  }, [readTokenAndConnectUser])

  useAppStateChange(readTokenAndConnectUser, () => void 0, [isLoggedIn])

  const value = useMemo(() => ({ isLoggedIn, setIsLoggedIn }), [isLoggedIn, setIsLoggedIn])

  if (loading) return null
  /**
   * warning: for a better data integrity, use setIsLoggedIn only from specific places
   * where it is needed:
   * - useLogoutRoutine: when logging out
   * - useLoginRoutine: when applying the logging in (signin or emailValidation)
   */
  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>
})

export function useLoginRoutine() {
  const { setIsLoggedIn } = useAuthContext()
  const resetContexts = useResetContexts()

  /**
   * Executes the minimal set of instructions required to proceed to the login
   * @param {SigninResponse} response
   * @param {LoginRoutineMethod} method The process that triggered the login routine
   */
  const loginRoutine = async (response: SigninResponse, method: LoginRoutineMethod) => {
    connectUserToBatchAndFirebase(response.accessToken)
    await saveRefreshToken(response.refreshToken)
    await storage.saveString('access_token', response.accessToken)
    analytics.logLogin({ method })
    setIsLoggedIn(true)
    resetContexts()
  }

  return loginRoutine
}

export function useLogoutRoutine(): () => Promise<void> {
  const queryClient = useQueryClient()
  const { setIsLoggedIn } = useAuthContext()

  return useCallback(async () => {
    try {
      BatchUser.editor().setIdentifier(null).save()
      analytics.logLogout()
      LoggedInQueryKeys.forEach((queryKey) => {
        queryClient.removeQueries(queryKey)
      })
      await storage.clear('access_token')
      await clearRefreshToken()
    } catch (err) {
      eventMonitoring.captureException(err)
    } finally {
      setIsLoggedIn(false)
    }
  }, [setIsLoggedIn])
}

// List of keys that are accessible only when logged in
// To clean when logging out
const LoggedInQueryKeys: QueryKeys[] = [
  QueryKeys.BOOKINGS,
  QueryKeys.FAVORITES,
  QueryKeys.FAVORITES_COUNT,
  QueryKeys.RECOMMENDATION_HITS,
  QueryKeys.RECOMMENDATION_OFFER_IDS,
  QueryKeys.REPORTED_OFFERS,
  QueryKeys.REPORT_OFFER_REASONS,
  QueryKeys.USER_PROFILE,
]
