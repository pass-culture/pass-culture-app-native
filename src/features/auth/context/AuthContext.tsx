import pick from 'lodash/pick'
import React, { memo, useCallback, useContext, useEffect, useMemo, useRef, useState } from 'react'
import { QueryObserverResult } from 'react-query'

import { api } from 'api/api'
import { UserProfileResponse } from 'api/gen'
import { useCookies } from 'features/cookies/helpers/useCookies'
import { navigateFromRef } from 'features/navigation/navigationRef'
// eslint-disable-next-line no-restricted-imports
import { amplitude } from 'libs/amplitude'
import { useAppStateChange } from 'libs/appState'
// eslint-disable-next-line no-restricted-imports
import { firebaseAnalytics } from 'libs/firebase/analytics'
import {
  computeTokenRemainingLifetimeInMs,
  getTokenStatus,
  getUserIdFromAccessToken,
} from 'libs/jwt'
import { getTokenExpirationDate } from 'libs/jwt/getTokenExpirationDate'
import { clearRefreshToken, getRefreshToken } from 'libs/keychain'
import { eventMonitoring } from 'libs/monitoring'
import { useNetInfoContext } from 'libs/network/NetInfoWrapper'
import { getAppVersion } from 'libs/packageJson'
import { QueryKeys } from 'libs/queryKeys'
import { BatchUser } from 'libs/react-native-batch'
import { usePersistQuery } from 'libs/react-query/usePersistQuery'
import { storage } from 'libs/storage'
import { getAge } from 'shared/user/getAge'

const MAX_AVERAGE_SESSION_DURATION_IN_MS = 60 * 60 * 1000
export interface IAuthContext {
  isLoggedIn: boolean
  setIsLoggedIn: (isLoggedIn: boolean) => void
  user?: UserProfileResponse
  refetchUser: () => Promise<QueryObserverResult<UserProfileResponse, unknown>>
  isUserLoading: boolean
}

const NAVIGATION_DELAY_FOR_EXPIRED_REFRESH_TOKEN_IN_MS = 1000

const navigateToLoginWithHelpMessage = () =>
  navigateFromRef('Login', { displayForcedLoginHelpMessage: true })

export const useConnectServicesRequiringUserId = (): ((accessToken: string | null) => void) => {
  const { setUserId: setUserIdToCookiesChoice } = useCookies()

  return useCallback(
    (accessToken) => {
      if (!accessToken) return

      const userId = getUserIdFromAccessToken(accessToken)
      if (userId) {
        BatchUser.editor().setIdentifier(userId.toString()).save()
        firebaseAnalytics.setUserId(userId)
        eventMonitoring.setUser({ id: userId.toString() })

        eventMonitoring.setExtras({
          accessTokenExpirationDate:
            getTokenExpirationDate(accessToken) ?? "can't get access token expiration date",
        })
        setUserIdToCookiesChoice(userId)
      }
    },
    [setUserIdToCookiesChoice]
  )
}

export const AuthContext = React.createContext<IAuthContext>({
  isLoggedIn: false,
  setIsLoggedIn: () => undefined,
  user: undefined,
  refetchUser: async () => ({} as QueryObserverResult<UserProfileResponse>),
  isUserLoading: false,
})

export function useAuthContext(): IAuthContext {
  return useContext(AuthContext)
}

export const AuthWrapper = memo(function AuthWrapper({
  children,
}: {
  children: React.JSX.Element
}) {
  const timeoutRef = useRef<NodeJS.Timeout>()
  const navigationTimeoutRef = useRef<NodeJS.Timeout>()
  const [loading, setLoading] = useState(true)
  const [isLoggedIn, setIsLoggedIn] = useState(false)
  const connectServicesRequiringUserId = useConnectServicesRequiringUserId()
  const {
    data: user,
    refetch: refetchUser,
    isLoading: isUserLoading,
  } = useUserProfileInfo(isLoggedIn)

  const readTokenAndConnectUser = useCallback(async () => {
    try {
      const refreshToken = await getRefreshToken()
      const accessToken = await storage.readString('access_token')
      const refreshTokenStatus = getTokenStatus(refreshToken)

      switch (refreshTokenStatus) {
        case 'unknown':
          setIsLoggedIn(false)
          return
        case 'expired':
          setIsLoggedIn(false)
          // We need to delay this navigation to avoid conflit between this navigation and the initial screen defined by react-navigation on app launch
          navigationTimeoutRef.current = setTimeout(async () => {
            navigateToLoginWithHelpMessage()
            await clearRefreshToken()
          }, NAVIGATION_DELAY_FOR_EXPIRED_REFRESH_TOKEN_IN_MS)
          return
        case 'valid':
          setIsLoggedIn(true)
          connectServicesRequiringUserId(accessToken)
          if (refreshToken) {
            eventMonitoring.setExtras({
              refreshTokenExpirationDate:
                getTokenExpirationDate(refreshToken) ?? "can't get refresh token expiration date",
            })
            const remainingLifetimeInMs = computeTokenRemainingLifetimeInMs(refreshToken)
            if (
              remainingLifetimeInMs &&
              remainingLifetimeInMs < MAX_AVERAGE_SESSION_DURATION_IN_MS
            ) {
              timeoutRef.current = globalThis.setTimeout(async () => {
                setIsLoggedIn(false)
                navigateToLoginWithHelpMessage()
                await clearRefreshToken()
              }, remainingLifetimeInMs)
            }
          }
          return
      }
    } catch (err) {
      eventMonitoring.captureException(err)
      setIsLoggedIn(false)
    } finally {
      setLoading(false)
    }
  }, [connectServicesRequiringUserId])

  useEffect(() => {
    readTokenAndConnectUser()

    return () => {
      if (timeoutRef.current) clearTimeout(timeoutRef.current)
      if (navigationTimeoutRef.current) clearTimeout(navigationTimeoutRef.current)
    }
  }, [readTokenAndConnectUser])

  useAppStateChange(readTokenAndConnectUser, () => void 0, [isLoggedIn])

  useEffect(() => {
    if (!user?.id) return

    const partialUser = pick(user, [
      'depositType',
      'eligibility',
      'eligibilityEndDatetime',
      'id',
      'isBeneficiary',
      'needsToFillCulturalSurvey',
    ])
    amplitude.setUserProperties({
      ...partialUser,
      ...(user.birthDate ? { age: getAge(user.birthDate) } : {}),
      status: user.status?.statusType, // eligible, beneficiaire, suspendu, etc
      appVersion: getAppVersion(),
    })
    amplitude.setUserId(user.id.toString())
  }, [user])

  const value = useMemo(
    () => ({ isLoggedIn, setIsLoggedIn, user, refetchUser, isUserLoading }),
    [isLoggedIn, setIsLoggedIn, user, refetchUser, isUserLoading]
  )

  if (loading) return null
  /**
   * warning: for a better data integrity, use setIsLoggedIn only from specific places
   * where it is needed:
   * - useLogoutRoutine: when logging out
   * - useLoginRoutine: when applying the logging in (signin or emailValidation)
   */
  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>
})

const STALE_TIME_USER_PROFILE = 5 * 60 * 1000
function useUserProfileInfo(isLoggedIn: boolean, options = {}) {
  const netInfo = useNetInfoContext()
  return usePersistQuery<UserProfileResponse>([QueryKeys.USER_PROFILE], () => api.getNativeV1Me(), {
    enabled: !!netInfo.isConnected && isLoggedIn,
    staleTime: STALE_TIME_USER_PROFILE,
    ...options,
  })
}
