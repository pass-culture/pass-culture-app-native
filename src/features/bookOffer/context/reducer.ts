export enum Step {
  DATE = 1,
  HOUR = 2,
  PRICE = 3,
  DUO = 4,
  CONFIRMATION = 5,
}

export const STEP_LABEL: Record<Step, string> = {
  [Step.DATE]: 'Date',
  [Step.HOUR]: 'Horaire',
  [Step.PRICE]: 'Prix',
  [Step.DUO]: 'Nombre de places',
  [Step.CONFIRMATION]: 'Confirmation',
}

export type BookingState = {
  offerId: number | undefined
  stockId: number | undefined
  step: Step
  quantity: 1 | 2 | undefined
  date: Date | undefined
  hour: string | undefined
}

export const initialBookingState: BookingState = {
  offerId: undefined,
  stockId: undefined,
  step: Step.DATE,
  quantity: undefined,
  date: undefined,
  hour: undefined,
}

export type Action =
  | { type: 'SET_OFFER_ID'; payload: number }
  | { type: 'VALIDATE_OPTIONS' }
  | { type: 'CHANGE_STEP'; payload: Step }
  | { type: 'CHANGE_OFFER'; payload: number }
  | { type: 'SELECT_DATE'; payload: Date }
  | { type: 'SELECT_HOUR'; payload: string }
  | { type: 'SELECT_STOCK'; payload: number }
  | { type: 'SELECT_QUANTITY'; payload: 1 | 2 }
  | { type: 'RESET' }
  | { type: 'RESET_HOUR' }
  | { type: 'RESET_STOCK' }
  | { type: 'RESET_QUANTITY' }

export const bookOfferReducer = (state: BookingState, action: Action): BookingState => {
  switch (action.type) {
    case 'RESET':
      return { ...initialBookingState, offerId: state.offerId }
    case 'SET_OFFER_ID':
      return { ...initialBookingState, offerId: action.payload }
    case 'VALIDATE_OPTIONS':
      return { ...state, step: Step.CONFIRMATION }
    case 'SELECT_STOCK':
      return { ...state, stockId: action.payload }
    case 'RESET_STOCK':
      return { ...state, stockId: undefined }
    case 'SELECT_QUANTITY':
      return { ...state, quantity: action.payload }
    case 'RESET_QUANTITY':
      return { ...state, quantity: undefined }
    case 'CHANGE_STEP':
      return { ...state, step: action.payload }
    case 'SELECT_DATE':
      return { ...state, date: action.payload }
    case 'SELECT_HOUR':
      return { ...state, hour: action.payload }
    case 'RESET_HOUR':
      return { ...state, hour: undefined }
    case 'CHANGE_OFFER':
      return { ...state, offerId: action.payload }
    default:
      return state
  }
}
