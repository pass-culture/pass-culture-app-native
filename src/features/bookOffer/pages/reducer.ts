export enum Step {
  DATE,
  HOUR,
  DUO,
  CONFIRMATION,
}

export type BookingState = {
  offerId: number | undefined
  stockId: number | undefined
  step: Step | undefined
  quantity: 1 | 2
}

export const initialBookingState: BookingState = {
  offerId: undefined,
  stockId: undefined,
  step: undefined,
  quantity: 1,
}

export type Action =
  | { type: 'INIT'; payload: Partial<BookingState> }
  | { type: 'VALIDATE_OPTIONS' }
  | { type: 'SELECT_QUANTITY'; payload: 1 | 2 }
  | { type: 'MODIFY_OPTIONS' }
  | { type: 'SELECT_STOCK'; payload: number }

export const searchReducer = (state: BookingState, action: Action): BookingState => {
  switch (action.type) {
    case 'INIT':
      return { ...initialBookingState, ...action.payload }
    case 'VALIDATE_OPTIONS':
      return { ...state, step: Step.CONFIRMATION }
    case 'SELECT_STOCK':
      return { ...state, stockId: action.payload }
    case 'SELECT_QUANTITY':
      return { ...state, quantity: action.payload }
    case 'MODIFY_OPTIONS':
      return { ...state, step: Step.DUO }
    default:
      return state
  }
}
