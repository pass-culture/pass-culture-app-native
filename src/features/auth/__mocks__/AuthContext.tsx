import { useAuthContext as actualUseAuthContext } from '../AuthContext'

export const setIsLoggedIn = jest.fn()

export const useAuthContext = jest.fn().mockReturnValue({
  setIsLoggedIn,
  isLoggedIn: false,
}) as jest.MockedFunction<typeof actualUseAuthContext>

export const signOut = jest.fn()

export const useLogoutRoutine = jest.fn(() => signOut)

export const loginRoutine = jest.fn()

export const useLoginRoutine = jest.fn(() => loginRoutine)
