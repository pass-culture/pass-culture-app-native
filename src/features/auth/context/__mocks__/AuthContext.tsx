import { beneficiaryUser } from 'fixtures/user'

import { useAuthContext as actualUseAuthContext } from '../AuthContext'

export const setIsLoggedIn = jest.fn()

export const useAuthContext = jest.fn().mockReturnValue({
  setIsLoggedIn,
  isLoggedIn: true,
  user: beneficiaryUser,
  isUserLoading: false,
}) as jest.MockedFunction<typeof actualUseAuthContext>
