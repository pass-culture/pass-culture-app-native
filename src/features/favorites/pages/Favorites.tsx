import { t } from '@lingui/macro'
import React from 'react'
import styled from 'styled-components/native'

import { useAuthContext } from 'features/auth/AuthContext'
import { FavoritesResults } from 'features/favorites/components/FavoritesResults'
import { NotConnectedFavorites } from 'features/favorites/components/NotConnectedFavorites'
import SvgPageHeader from 'ui/components/headers/SvgPageHeader'
import { useKeyboardAdjust } from 'ui/components/keyboard/useKeyboardAdjust'

export const Favorites: React.FC = () => {
  useKeyboardAdjust()
  const { isLoggedIn } = useAuthContext()

  if (!isLoggedIn) return <NotConnectedFavorites />

  return (
    <Container>
      <SvgPageHeader title={t`Mes favoris`} />
      <FavoritesResults />
    </Container>
  )
}

const Container = styled.View({ flex: 1 })
