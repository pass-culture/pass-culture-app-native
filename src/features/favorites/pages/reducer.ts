import { FavoriteSortBy } from 'features/favorites/pages/FavoritesSorts'

type FavoritesState = {
  sortBy: FavoriteSortBy
}

export type FavoritesContext = FavoritesState & {
  dispatch: (action: Action) => void
}

export const initialFavoritesState: FavoritesState = {
  sortBy: 'RECENTLY_ADDED',
}

type Action = { type: 'INIT' } | { type: 'SET_SORT_BY'; payload: FavoriteSortBy }

export const favoritesReducer = (state: FavoritesState, action: Action): FavoritesState => {
  switch (action.type) {
    case 'INIT':
      return initialFavoritesState
    case 'SET_SORT_BY':
      return { ...state, sortBy: action.payload }
    default:
      return state
  }
}
