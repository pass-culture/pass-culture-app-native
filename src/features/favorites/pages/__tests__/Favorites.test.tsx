import { cleanup, render } from '@testing-library/react-native'
import React from 'react'

import { useAuthContext } from 'features/auth/AuthContext'
import { reactQueryProviderHOC } from 'tests/reactQueryProviderHOC'

import { Favorites } from '../Favorites'
import { initialFavoritesState } from '../reducer'

const mockSearchState = initialFavoritesState
const mockDispatch = jest.fn()
jest.mock('features/favorites/pages/FavoritesWrapper', () => ({
  useFavoritesState: () => ({
    ...mockSearchState,
    dispatch: mockDispatch,
  }),
}))
jest.mock('features/auth/AuthContext')
const mockUseAuthContext = useAuthContext as jest.MockedFunction<typeof useAuthContext>

describe('Favorites component', () => {
  afterEach(async () => {
    jest.clearAllMocks()
    await cleanup()
  })

  it('should render correctly', () => {
    const { toJSON } = renderFavorites({ isLoggedIn: true })
    expect(toJSON()).toMatchSnapshot()
  })

  it('should show non connected page when not logged in', () => {
    const { getByText } = renderFavorites({ isLoggedIn: false })
    expect(getByText('Connecte-toi pour profiter de cette fonctionnalitÃ© !')).toBeTruthy()
  })

  it('should show loading when not logged in', () => {
    renderFavorites({ isLoggedIn: false })
    expect(mockDispatch).not.toBeCalledWith({ type: 'SHOW_RESULTS', payload: true })
  })
})

function renderFavorites({ isLoggedIn }: { isLoggedIn: boolean }) {
  const setIsLoggedIn = jest.fn()
  mockUseAuthContext.mockImplementation(() => ({ isLoggedIn, setIsLoggedIn }))
  return render(reactQueryProviderHOC(<Favorites />))
}
