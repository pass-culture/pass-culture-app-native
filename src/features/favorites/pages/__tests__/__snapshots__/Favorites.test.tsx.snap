// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites component should render correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "maxHeight": 76,
          "overflow": "hidden",
          "position": "relative",
        },
      ]
    }
  >
    <View
      height={280}
      width={751}
    >
      <Text>
        undefined-SVG-Mock
      </Text>
    </View>
    <Text
      style={
        Array [
          Object {
            "bottom": 12,
            "color": "#ffffff",
            "fontFamily": "Montserrat-Medium",
            "fontSize": 18,
            "lineHeight": 22,
            "position": "absolute",
          },
        ]
      }
    >
      Mes favoris
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "height": "100%",
        },
      ]
    }
    testID="FavoritesResultsPlaceHolder"
  >
    <RCTScrollView
      ItemSeparatorComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "displayName": "Styled(View)",
          "inlineStyle": InlineStyle {
            "rules": Array [
              "height: 2px;",
              "background-color: #f5f5f5;",
              "margin-horizontal: 24px;",
              "margin-vertical: 16px;",
            ],
          },
          "render": [Function],
          "shouldForwardProp": undefined,
          "styledComponentId": "StyledNativeComponent",
          "target": [Function],
          "withComponent": [Function],
        }
      }
      ListFooterComponent={<Styled(View) />}
      ListHeaderComponent={<NumberOfResultsPlaceholder />}
      contentContainerStyle={
        Object {
          "flexGrow": 1,
          "paddingBottom": 80,
        }
      }
      data={
        Array [
          Object {
            "key": "0",
          },
          Object {
            "key": "1",
          },
          Object {
            "key": "2",
          },
          Object {
            "key": "3",
          },
          Object {
            "key": "4",
          },
          Object {
            "key": "5",
          },
          Object {
            "key": "6",
          },
          Object {
            "key": "7",
          },
          Object {
            "key": "8",
          },
          Object {
            "key": "9",
          },
          Object {
            "key": "10",
          },
          Object {
            "key": "11",
          },
          Object {
            "key": "12",
          },
          Object {
            "key": "13",
          },
          Object {
            "key": "14",
          },
          Object {
            "key": "15",
          },
          Object {
            "key": "16",
          },
          Object {
            "key": "17",
          },
          Object {
            "key": "18",
          },
          Object {
            "key": "19",
          },
        ]
      }
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEnabled={false}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
        >
          <View
            style={
              Array [
                Object {
                  "marginHorizontal": 24,
                },
              ]
            }
          >
            <View
              numberOfSpaces={7}
              style={
                Array [
                  Object {
                    "height": 28,
                  },
                ]
              }
            />
            <View
              numberOfSpaces={5}
              style={
                Array [
                  Object {
                    "height": 20,
                  },
                ]
              }
            />
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Array [
                Object {
                  "marginHorizontal": 24,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                  },
                ]
              }
            />
          </View>
          <View
            highlighted={false}
            leadingItem={
              Object {
                "key": "0",
              }
            }
            style={
              Array [
                Object {
                  "backgroundColor": "#f5f5f5",
                  "height": 2,
                  "marginHorizontal": 24,
                  "marginVertical": 16,
                },
              ]
            }
          />
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Array [
                Object {
                  "marginHorizontal": 24,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                  },
                ]
              }
            />
          </View>
          <View
            highlighted={false}
            leadingItem={
              Object {
                "key": "1",
              }
            }
            style={
              Array [
                Object {
                  "backgroundColor": "#f5f5f5",
                  "height": 2,
                  "marginHorizontal": 24,
                  "marginVertical": 16,
                },
              ]
            }
          />
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Array [
                Object {
                  "marginHorizontal": 24,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                  },
                ]
              }
            />
          </View>
          <View
            highlighted={false}
            leadingItem={
              Object {
                "key": "2",
              }
            }
            style={
              Array [
                Object {
                  "backgroundColor": "#f5f5f5",
                  "height": 2,
                  "marginHorizontal": 24,
                  "marginVertical": 16,
                },
              ]
            }
          />
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Array [
                Object {
                  "marginHorizontal": 24,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                  },
                ]
              }
            />
          </View>
          <View
            highlighted={false}
            leadingItem={
              Object {
                "key": "3",
              }
            }
            style={
              Array [
                Object {
                  "backgroundColor": "#f5f5f5",
                  "height": 2,
                  "marginHorizontal": 24,
                  "marginVertical": 16,
                },
              ]
            }
          />
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Array [
                Object {
                  "marginHorizontal": 24,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                  },
                ]
              }
            />
          </View>
          <View
            highlighted={false}
            leadingItem={
              Object {
                "key": "4",
              }
            }
            style={
              Array [
                Object {
                  "backgroundColor": "#f5f5f5",
                  "height": 2,
                  "marginHorizontal": 24,
                  "marginVertical": 16,
                },
              ]
            }
          />
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Array [
                Object {
                  "marginHorizontal": 24,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                  },
                ]
              }
            />
          </View>
          <View
            highlighted={false}
            leadingItem={
              Object {
                "key": "5",
              }
            }
            style={
              Array [
                Object {
                  "backgroundColor": "#f5f5f5",
                  "height": 2,
                  "marginHorizontal": 24,
                  "marginVertical": 16,
                },
              ]
            }
          />
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Array [
                Object {
                  "marginHorizontal": 24,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                  },
                ]
              }
            />
          </View>
          <View
            highlighted={false}
            leadingItem={
              Object {
                "key": "6",
              }
            }
            style={
              Array [
                Object {
                  "backgroundColor": "#f5f5f5",
                  "height": 2,
                  "marginHorizontal": 24,
                  "marginVertical": 16,
                },
              ]
            }
          />
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Array [
                Object {
                  "marginHorizontal": 24,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                  },
                ]
              }
            />
          </View>
          <View
            highlighted={false}
            leadingItem={
              Object {
                "key": "7",
              }
            }
            style={
              Array [
                Object {
                  "backgroundColor": "#f5f5f5",
                  "height": 2,
                  "marginHorizontal": 24,
                  "marginVertical": 16,
                },
              ]
            }
          />
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Array [
                Object {
                  "marginHorizontal": 24,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                  },
                ]
              }
            />
          </View>
          <View
            highlighted={false}
            leadingItem={
              Object {
                "key": "8",
              }
            }
            style={
              Array [
                Object {
                  "backgroundColor": "#f5f5f5",
                  "height": 2,
                  "marginHorizontal": 24,
                  "marginVertical": 16,
                },
              ]
            }
          />
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Array [
                Object {
                  "marginHorizontal": 24,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                  },
                ]
              }
            />
          </View>
          <View
            highlighted={false}
            leadingItem={
              Object {
                "key": "9",
              }
            }
            style={
              Array [
                Object {
                  "backgroundColor": "#f5f5f5",
                  "height": 2,
                  "marginHorizontal": 24,
                  "marginVertical": 16,
                },
              ]
            }
          />
        </View>
        <View
          style={
            Object {
              "height": 0,
            }
          }
        />
        <View
          onLayout={[Function]}
        >
          <View
            style={
              Array [
                Object {
                  "height": 0,
                },
              ]
            }
          />
        </View>
      </View>
    </RCTScrollView>
  </View>
  <View
    style={
      Array [
        Object {
          "alignSelf": "center",
          "bottom": 88,
          "position": "absolute",
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
          "overflow": "hidden",
        }
      }
      testID="SortButton"
    >
      <BVLinearGradient
        colors={
          Array [
            4290717535,
            4284091776,
          ]
        }
        endPoint={
          Object {
            "x": 1,
            "y": 0,
          }
        }
        locations={null}
        startPoint={
          Object {
            "x": 0,
            "y": 0,
          }
        }
        style={
          Array [
            Object {
              "alignItems": "center",
              "borderRadius": 24,
              "flexDirection": "row",
              "height": 40,
              "overflow": "hidden",
              "paddingHorizontal": 12,
            },
          ]
        }
      >
        <View
          fill="#ffffff"
          height={32}
          width={32}
        >
          <Text>
            undefined-SVG-Mock
          </Text>
        </View>
        <View
          numberOfSpaces={1}
          style={
            Array [
              Object {
                "width": 4,
              },
            ]
          }
        />
        <Text
          style={
            Array [
              Object {
                "color": "#ffffff",
                "fontFamily": "Montserrat-Bold",
                "fontSize": 15,
                "lineHeight": 20,
              },
            ]
          }
        >
          Trier
        </Text>
        <View
          numberOfSpaces={2}
          style={
            Array [
              Object {
                "width": 8,
              },
            ]
          }
        />
      </BVLinearGradient>
    </View>
    <View
      customHeight={8}
      style={
        Array [
          Object {
            "height": 8,
          },
        ]
      }
    />
  </View>
</View>
`;
